### ISSUE TITLE
How to register a WCF ServiceHost in a self-hosted environment

### ISSUE THREAD

**Issue by solomon87** [simpleinjector/SimpleInjector#594]
I'm trying to figure out how to register my ServiceHost with SimpleInjector so that services are properly injected into its constructor. In addition I want to make sure that I am scoping the services correctly. Looking at the WCF Integration documentation talks about using the supplied SimpleInjectorServiceHostFactory but that seems to only apply in IIS hosted situations. In addition, I'm not sure the way I'm scoping my ServiceHost instance is correct given its context within the windows service MyServiceHost. Any help with this would be greatly appreciated.

```csharp
public partial class MyServiceHost : ServiceBase
{
    private ServiceHost _host;
    private Scope _asyncScope;

    public MyServiceHost ()
    {
        _asyncScope = AsyncScopedLifestyle.BeginScope(SimpleServiceProvider.Container);
        _host = _asyncScope.Container.GetInstance<ServiceHost>();

        InitializeComponent();
    }

    protected override void OnStart(string[] args)
    {
        try
        {
            _host.Open();
        }
        catch (Exception ex)
        {
            throw;
        }
    }

    protected override void OnStop()
    {
        _host.Close();
        _asyncScope.Dispose();
    }
}

[ServiceBehavior(
    
    ConcurrencyMode = ConcurrencyMode.Single,
    InstanceContextMode = InstanceContextMode.PerCall,
    MaxItemsInObjectGraph = int.MaxValue)]
public class MyService : InfrastructureServiceBase, IMyService
{

    private readonly TaskManager _taskManager;

    public MyService(TaskManager taskManager)
    {
        _taskManager = taskManager;
    }

    // Public Methods
    public IInstruction ExecuteTwoWayService(MyRequest request)
    {
        try
        {
            return _taskManager.ExecuteTwoWayService(request);
        }
        catch (AggregateException aex)
        {
            HandleAggregateException(aex, "MyService", "ExecuteThisTwoWayService");
        }

        return null;
    }

}

public static class SimpleServiceProvider
{
    private static bool _isBootstrapped = false;

    public static Container Container { get; }

    static SimpleServiceProvider()
    {
        Container = new Container();
    }

    public static object GetInstance(Type serviceType)
    {
        return Container.GetInstance(serviceType);
    }

    public static TService GetInstance<TService>()
    {
        return (TService) GetInstance(typeof(TService));
    }

    public static void Bootstrap()
    {
        if (_isBootstrapped) throw new InvalidOperationException("SimpleServiceProvider is already bootstrapped.");

        Container.Options.DefaultLifestyle = Lifestyle.Scoped;
        Container.Options.DefaultScopedLifestyle = new AsyncScopedLifestyle();

        Uri hostUri = ServiceConfigurationFactory.GetServiceUriByDefaultLocation("IMyService");

        // Register Factories
        Container.Register<IServiceHostFactory<IMyService>>(() => new ServiceHostFactory<IMyService>(hostUri), Lifestyle.Singleton);

        // Register WCF services.
//            Container.RegisterWcfServices(Assembly.GetExecutingAssembly());
        Container.Register<ServiceHost>(() => Container.GetInstance<IServiceHostFactory<IMyService>>().GetService<MyService>());

//            SimpleInjectorServiceHostFactory.SetContainer(_container);

        Container.Verify();
        _isBootstrapped = true;
    }
}
```

**Comment by solomon87** [comment]
Fix code

**Comment by dotnetjunkie** [comment]
> Looking at the WCF Integration documentation talks about using the supplied SimpleInjectorServiceHostFactory but that seems to only apply in IIS hosted situations.

The [second part of the WCF Integration Guide](https://simpleinjector.org/wcf#was-hosting-and-non-http-activation) talks about WAS Hosting and Non-HTTP Activation, which is what you are talking about. If you follow that guide, you should be able to get everything running in no time in an non-IIS environment.

> In addition, I'm not sure the way I'm scoping my ServiceHost instance is correct given its context within the windows service MyServiceHost.

When you follow the guidance, you create a custom `SimpleInjectorServiceHostFactory` implementation that returns a `SimpleInjectorServiceHost`. `SimpleInjectorServiceHost` ensures correct scoping around the WCF operation.

**Comment by solomon87** [comment]
The very last part is the part I'm still having trouble with. Our implementation doesn't use .svc files. But with some digging I believe I found how to specify the factory without that file.

```
<system.serviceModel>
    <serviceHostingEnvironment>
      <serviceActivations>
        <add service="MyTest.Service" factory="MyTest.MyServiceHostFactory" relativeAddress="/Hello.svc" />
      </serviceActivations>
    </serviceHostingEnvironment>
  </system.serviceModel>
```

Could you add this to the documentation? Also, once I have all this setup, do I simply call into my static Bootstrap(er) class from the Program.Main function?

**Comment by dotnetjunkie** [comment]
> Could you add this to the documentation?

Good point. I'll put it on [the documentation backlog](https://github.com/simpleinjector/documentation/issues).

> Also, how does my Program.Main function play into all this? Will .NET somehow bootstrap everything through the factory without any additional code there?

If you follow the documentation [example](https://simpleinjector.org/wcf#was-hosting-and-non-http-activation), when WCF calls your `MyServiceHostFactory.CreateServiceHost`, that method will get the `Bootstrapper.Container` field. If that is the first time that the `Bootstrapper` class is called, it will cause its static constructor to run, which will initialize the `Container` field.

> NOTE: To be pedantic, the `Bootstrapper`'s static constructor will run when the `CreateServiceHost` is JITted, just before its code will execute.

In other words, you don't have to do anything else in order to make sure the system is bootstrapped.

Downside, however, of static constructors is that when bootstrapping the container fails for some reason, e.g. because there are diagnostic errors, the thrown exception is wrapped in a different exception by the CLR, which might make it a bit harder to directly see what's happening.

**Comment by dotnetjunkie** [comment]
>  won't this create a single instance for the life of the Windows Service? 

Yes, there will be a single `ServiceHost` instance per WCF service class. This is by design; WCF will call `CreateServiceHost` once per WCF service class.

The the `SimpleInjectorServiceHost` is thread-safe and the underlying `SimpleInjectorInstanceProvider` will create a new WCF service instance on each call to its `GetInstance` method.

**Comment by solomon87** [comment]
Okay. My understanding was that one ServiceHost should be created per call (operation). It might have been this Tip that gave me the impression. Is this not the case?

> TIP: Use InstanceContextMode.PerCall for all WCF services. This prevents any hard to detect problems caused by WCF services outliving a single request.

In addition, if I spin off any non-awaited tasks as a result of my service call which _will_ outlive the original request, that should all still be no problem?

**Comment by dotnetjunkie** [comment]
> In addition, if I spin off any non-awaited tasks as a result of my service call which will outlive the original request, that should all still be no problem?

In that case you are not running code *asynchronously*—you are running your code in *parallel*. Would that be a problem? Yes! That would absolutely cause lots of pain. 

The first thing you'll notice is that any Scoped dependencies will get disposed when the request ends, which might be long before your background operation ends. But things get worse, pretty quickly.

You should not spin of any background threads from application code where you pass on components and dependencies to background threads, because that could easily lead to all kinds of race conditions. Just imagine having some object graph that contains an Entity Framework `DbContext`. A `DbContext` is *not* thread-safe, and can cause [all sorts of troubles](https://stackoverflow.com/questions/3266295/net-entity-framework-and-transactions#3266481) when used in parallel.

The [documentaiton](https://simpleinjector.org/howto#work-with-dependency-injection-in-multi-threaded-applications) actually goes into more detail on how to approach this. In summary, the advice is:

> In a multi-threaded application, each thread should get its own object graph. This means that you should typically call `GetInstance<T>()` once at the beginning of the thread’s execution to get the root object for processing that thread (or request). 

> The advice of building a new object graph (calling `GetInstance<T>()`) at the beginning of a thread, also holds when manually starting a new (background) thread. Although you can pass on data to other threads, you should not pass on container controlled dependencies to other threads. On each new thread, you should ask the container again for the dependencies. When you start passing dependencies from one thread to the other, those parts of the code have to know whether it is safe to pass those dependencies on.

Your case—starting a background operation using a `Task` by not awaiting it—is very similar to described by the document; the advice is identical:

> hide the logic for asynchronicity behind an abstraction; a proxy [...] and [place] this proxy inside the Composition Root [to] prevent the application code [from taking] a dependency on the container itself.

**Comment by solomon87** [comment]
Ah this is exactly what I'm looking for. Thanks!
